### ISSUE TITLE
t.p.t.ThreadPool.dispatch should be deprecated

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2684]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)| @bitglue reported|
|-|-|
|Trac ID|trac#2684|
|Type|enhancement|
|Created|2007-05-30 17:06:57Z|

exarkun : indigo: It should be deprecated.
exarkun : (it should have been deprecated 5 years ago actually)

<details><summary>Searchable metadata</summary>

```
trac-id__2684 2684
type__enhancement enhancement
reporter__indigo indigo
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1180544817000000 1180544817000000
changetime__1190886153000000 1190886153000000
version__None None
owner__ 
cc__therve cc__mwh
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

The work is done in dispatch-deprecation-2684, wait for the review stack to be less impressive.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

The unit tests for the deprecated functions shouldn't be removed until the deprecated functions are removed.

Also, I'm not sure we should sneakily change ThreadPool to be new-style instead of classic. :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

I added the test back, and removed the sneaky change :).

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_7"></a>|@mwhudson commented|
|-|-|

Some pickiness:

 - why change the import style at the top of the file? (for launchpad we have a rule that imports must be in alphabetical order, which I like)
 - in the docstring for ThreadPool you should say "callInThread() and stop()" rather than "callInThread and stop()"
 - also add ()s to the warning messages
 - you should explain what the deprecations are in favour of in the docstrings of the deprecated things.
 - "Test threadpool pickling/unpickling." isn't much of a docstring, I'm afraid.  "threadpools can be pickled and unpickled, which should preserve the number of threads and other parameters"?
 - there's a typo in the docstring for _threadpoolTest (dispath), though I'm not sure this docstring is long enough and calling the argument 'method' is unhelpful
 - "Test a basic acquire in a thread on an existing lock."  I have no idea what this is testing from this docstring (nor from reading the test either, which isn't your fault).
 - there's a superfluous blank line at the end of test_threadpool

Nothing major here though.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)|@mwhudson set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [21317]) Apply review comments.

Refs [#2684](https://github.com/twisted/twisted/issues/2684)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [21318]) Fix a typo and warning suppression.

Refs [#2684](https://github.com/twisted/twisted/issues/2684)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @mwhudson|
|-|-|

Replying to [mwh](#note_7):
> Some pickiness:
>
>  - why change the import style at the top of the file? (for launchpad we have a rule that imports must be in alphabetical order, which I like)

I didn't remember, so I reverted that. That should eventually be specified.

>  - in the docstring for ThreadPool you should say "callInThread() and stop()" rather than "callInThread and stop()"
>  - also add ()s to the warning messages
>  - you should explain what the deprecations are in favour of in the docstrings of the deprecated things.
>  - "Test threadpool pickling/unpickling." isn't much of a docstring, I'm afraid.  "threadpools can be pickled and unpickled, which should preserve the number of threads and other parameters"?
>  - there's a typo in the docstring for _threadpoolTest (dispath), though I'm not sure this docstring is long enough and calling the argument 'method' is unhelpful
>  - "Test a basic acquire in a thread on an existing lock."  I have no idea what this is testing from this docstring (nor from reading the test either, which isn't your fault).

Done.

>  - there's a superfluous blank line at the end of test_threadpool

I like that... Again, we have to update the coding standard document.

> Nothing major here though.

Thanks a lot for your review :).

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_12"></a>|@mwhudson set owner to @therve|
|-|-|

Only one comment:

 - four blank lines after ThreadPoolTestCase seems excessive :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set status to `closed`|
|-|-|

(In [21320]) Merge dispatch-deprecation-2684

Author: therve
Reviewers: mwh, exarkun
Fixes [#2684](https://github.com/twisted/twisted/issues/2684)

Deprecate the dispatch and dispatchWithCallback methods of
twisted.python.threadpool.ThreadPool, in favor of the callInThread method on
the same object and of twisted.internet.threads.deferToThread.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
